starknet/src/utils/eip712.cairo

```diff
@@ -0,0 +1,203 @@
+    encoded_data.append(metadata_URI.keccak_struct_hash());
+    encoded_data.append(salt);
+    let message_hash = keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse();
+    hash_typed_data(domain_hash, message_hash)
+}
+
+fn get_vote_digest(
+    domain_hash: u256,
+    space: ContractAddress,
+    voter: EthAddress,
+    proposal_id: u256,
+    choice: Choice,
+    user_voting_strategies: Span<IndexedStrategy>,
+    metadata_URI: Span<felt252>,
+) -> u256 {
+    let mut encoded_data = ArrayTrait::<u256>::new();
+    encoded_data.append(u256 { low: VOTE_TYPEHASH_LOW, high: VOTE_TYPEHASH_HIGH });
+    encoded_data.append(get_contract_address().into());
+    encoded_data.append(space.into());
+    encoded_data.append(voter.into());
+    encoded_data.append(proposal_id);
+    encoded_data.append(choice.into());
+    encoded_data.append(user_voting_strategies.keccak_struct_hash());
+    encoded_data.append(metadata_URI.keccak_struct_hash());
+    let message_hash = keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse();
+    hash_typed_data(domain_hash, message_hash)
+}
+
+fn get_update_proposal_digest(
+    domain_hash: u256,
+    space: ContractAddress,
+    author: EthAddress,
+    proposal_id: u256,
+    execution_strategy: @Strategy,
+    metadata_URI: Span<felt252>,
+    salt: u256
+) -> u256 {
+    let mut encoded_data = ArrayTrait::<u256>::new();
+    encoded_data
+        .append(u256 { low: UPDATE_PROPOSAL_TYPEHASH_LOW, high: UPDATE_PROPOSAL_TYPEHASH_HIGH });
+    encoded_data.append(get_contract_address().into());
+    encoded_data.append(space.into());
+    encoded_data.append(author.into());
+    encoded_data.append(proposal_id);
+    encoded_data.append(execution_strategy.keccak_struct_hash());
+    encoded_data.append(metadata_URI.keccak_struct_hash());
+    encoded_data.append(salt);
+    let message_hash = keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse();
+    hash_typed_data(domain_hash, message_hash)
+}
+
+fn get_domain_hash() -> u256 {
+    let mut encoded_data = ArrayTrait::<u256>::new();
+    encoded_data.append(u256 { low: DOMAIN_TYPEHASH_LOW, high: DOMAIN_TYPEHASH_HIGH });
+    encoded_data.append(get_tx_info().unbox().chain_id.into());
+    keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse()
+}
+
+fn hash_typed_data(domain_hash: u256, message_hash: u256) -> u256 {
+    let mut encoded_data = ArrayTrait::<u256>::new();
+    encoded_data.append(domain_hash);
+    encoded_data.append(message_hash);
+    let encoded_data = _add_prefix_array(encoded_data, ETHEREUM_PREFIX);
+    let (mut u64_arr, overflow) = into_le_u64_array(encoded_data);
+    keccak::cairo_keccak(ref u64_arr, overflow, 2).byte_reverse()
+}
+
+// Prefixes a 16 bit prefix to an array of 256 bit values.
+fn _add_prefix_array(input: Array<u256>, mut prefix: u128) -> Array<u256> {
+    let mut out = ArrayTrait::<u256>::new();
+    let mut i = 0_usize;
+    loop {
+        if i >= input.len() {
```

Comment:
issue here is that we want to do something different on the last element of the array. so kind of need to track the index instead of just popping each element off. 

using pop front then also tracking an index is probably less efficient than just tracking the index in the loop

Reply:
yep moved over